src/MediatR/Mediator.cs

```diff
@@ -73,20 +88,65 @@ public void Publish(INotification notification)
 
         public Task PublishAsync(IAsyncNotification notification)
         {
-            var notificationHandlers = GetNotificationHandlers(notification)
+            var notificationHandlers = GetNotificationHandlers(notification);
+
+            if (_publishOption == PublishAsyncOptions.Parallel)
+            {
+                return PublishParallelAsync(notification, notificationHandlers);
+            }
+            else
+            {
+                return PublishSequentialAsync(notification, notificationHandlers);
+            }
+        }
+
+        public Task PublishAsync(ICancellableAsyncNotification notification, CancellationToken cancellationToken)
+        {
+            var notificationHandlers = GetNotificationHandlers(notification);
+
+            if (_publishOption == PublishAsyncOptions.Parallel)
+            {
+                return PublishParallelAsync(notification, cancellationToken, notificationHandlers);
+            }
+            else
+            {
+                return PublishSequentialAsync(notification, cancellationToken, notificationHandlers);
+            }
+        }
+
+        private async Task PublishSequentialAsync(IAsyncNotification notification, IEnumerable<AsyncNotificationHandlerWrapper> notificationHandlers)
+        {
+            foreach (var handler in notificationHandlers)
+            {
+                await handler.Handle(notification);
+            }
+        }
+
+        private Task PublishParallelAsync(IAsyncNotification notification, IEnumerable<AsyncNotificationHandlerWrapper> notificationHandlers)
+        {
+            var tasks = notificationHandlers
                 .Select(handler => handler.Handle(notification))
                 .ToArray();
 
-            return Task.WhenAll(notificationHandlers);
+            return Task.WhenAll(tasks);
         }
 
-        public Task PublishAsync(ICancellableAsyncNotification notification, CancellationToken cancellationToken)
+        private Task PublishParallelAsync(ICancellableAsyncNotification notification, CancellationToken cancellationToken,
+            IEnumerable<CancellableAsyncNotificationHandlerWrapper> notificationHandlers)
         {
-            var notificationHandlers = GetNotificationHandlers(notification)
+            var tasks = notificationHandlers
                 .Select(handler => handler.Handle(notification, cancellationToken))
                 .ToArray();

```

Comment:
False or true, anyway let's be more friendly, right? I'm basing my statement on my experience that calling ToArray() kick-offs these tasks before they're reaching Task.WhenAll() what doesn't help much during debugging. Thanks for pointing out to the source code. But still I'd pass IEnumerable directly and let TPL do its work.